lib/bundler/resolver.rb

```diff
@@ -146,7 +146,43 @@ def search_for(dependency)
           @gem_version_promoter.sort_versions(dependency, spec_groups)
         end
       end
-      search.select {|sg| sg.for?(platform) }.each {|sg| sg.activate_platform!(platform) }
+      selected_sgs = []
+      search.each do |sg|
+        next unless sg.for?(platform)
+        spec_platform = sg.spec(platform).platform
+        if spec_platform && spec_platform != Gem::Platform::RUBY
+          # Add a spec group for "non platform specific spec" as the fallback
+          # spec group.
+          sg_ruby = SpecGroup.new(sg.all_specs)
+          sg_ruby.ignores_bundler_dependencies = sg.ignores_bundler_dependencies
+          if sg_ruby.for?(Gem::Platform::RUBY)
+            sg_ruby.activate_platform!(Gem::Platform::RUBY)
+            selected_sgs << sg_ruby
+          end
+        end
+        if platform != Gem::Platform::RUBY
+          # Add a spec group for ["non platform specific spec", "platform
+          # specific spec"] to resolve a spec for multiple platforms.
+          # If Gemfile.lock file has the following entries, we need a ffi-1.9.14
+          # spec group for ["ruby", "x86-mingw32"].
+          #
+          # GEM
+          #   specs:
+          #     ffi (1.9.14)
+          #     ffi (1.9.14-x86-mingw32)
+          sg_ruby_platform = SpecGroup.new(sg.all_specs)
+          sg_ruby_platform.ignores_bundler_dependencies =
+            sg.ignores_bundler_dependencies
+          if sg_ruby_platform.for?(Gem::Platform::RUBY)
+            sg_ruby_platform.activate_platform!(Gem::Platform::RUBY)
+            sg_ruby_platform.activate_platform!(platform)
+            selected_sgs << sg_ruby_platform
+          end
+        end
+        sg.activate_platform!(platform)
+        selected_sgs << sg
+      end
+      selected_sgs
```

Comment:
This change, `requirement_satisfied_by?` change and `self.platform_sort_key` change is the one of the most important changes for this improvement.